src/Network/OAuth/OAuth2/Internal.hs

```diff
@@ -63,12 +67,62 @@ instance FromJSON OAuth2Token where
 instance ToJSON OAuth2Token where
     toEncoding = genericToEncoding defaultOptions { fieldLabelModifier = camelTo2 '_' }
 
+-- | Possible OAuth error types
+-- | Not all of them can appear everywhere, but this is good enough for now.
```

Comment:
```haskell
{-# LANGUAGE DeriveGeneric              #-}

module Network.OAuth.OAuth2.TokenRequest where

import           Data.Aeson
import           Data.Aeson.Types
import           GHC.Generics

instance FromJSON Errors where
  parseJSON = genericParseJSON defaultOptions { constructorTagModifier = camelTo2 '_', allNullaryToStringTag = True }
instance ToJSON Errors where
  toEncoding = genericToEncoding defaultOptions { constructorTagModifier = camelTo2 '_', allNullaryToStringTag = True }

-- Token Error Responses https://tools.ietf.org/html/rfc6749#section-5.2
data Errors =
    InvalidRequest
  | InvalidClient
  | InvalidGrant
  | UnauthorizedClient
  | UnsupportedGrantType
  | InvalidScope
  deriving (Show, Eq, Generic)
```

The two sets share some similarities, but not all - not sure how to model that in haskell. i figured this way works decently well.

I'll code some more the day after tomorrow, then I'll push my suggestions.